src/libxtp/gaussian_quadrature.cc

```diff
@@ -0,0 +1,225 @@
+    : _energies(energies), _Mmn(Mmn) {}
+
+void GaussianQuadrature::configure(options opt, const RPA& rpa) {
+  _opt = opt;
+  if (_opt.quadrature_scheme == "laguerre") {
+    Gauss_Laguerre_Quadrature_Constants glqc;
+    _quadpoints = glqc.getPoints(_opt.order);
+    _quadadaptedweights = glqc.getAdaptedWeights(_opt.order);
+    CalcDielInvVector(rpa);
+  } else if (_opt.quadrature_scheme == "legendre") {
+    Gauss_Legendre_Quadrature_Constants glqc;
+    _quadpoints = glqc.getPoints(_opt.order);
+    _quadadaptedweights = glqc.getAdaptedWeights(_opt.order);
+    CalcDielInvVector(rpa);
+  } else if (_opt.quadrature_scheme == "modified_legendre") {
+    Gauss_Legendre_Quadrature_Constants glqc;
+    _quadpoints = glqc.getPoints(_opt.order);
+    _quadadaptedweights = glqc.getAdaptedWeights(_opt.order);
+    CalcDielInvVector(rpa);
+  } else if (_opt.quadrature_scheme == "hermite") {
+    Gauss_Hermite_Quadrature_Constants glqc;
+    _quadpoints = glqc.getPoints(_opt.order);
+    _quadadaptedweights = glqc.getAdaptedWeights(_opt.order);
+    CalcDielInvVector(rpa);
+  } else {
+    std::cout << "There no such a thing as the integration scheme you asked"
+              << std::endl;
+  }
+}
+
+// This function calculates and stores inverses of the microscopic dielectric
+// matrix in a matrix vector
+void GaussianQuadrature::CalcDielInvVector(const RPA& rpa) {
+  _dielinv_matrices_r.resize(_opt.order);
+  double halfpi = std::acos(0.0);
+#pragma openmp parallel schedule(guided)
+  for (Index j = 0; j < _opt.order; j++) {
+    if (_opt.quadrature_scheme == "legendre") {
+      double newpoint = std::tan(halfpi * _quadpoints(j));
+      Eigen::MatrixXcd eps_inv_j =
+          rpa.calculate_epsilon_complex(0.0, newpoint).inverse();
+      eps_inv_j.diagonal().array() -= 1.0;
+      _dielinv_matrices_r[j] = -eps_inv_j;
+    } else if (_opt.quadrature_scheme == "modified_legendre") {
+      double exponent = (1.0 + _quadpoints(j)) / (1.0 - _quadpoints(j));
+      double newpoint = std::pow(0.5, exponent);
+      Eigen::MatrixXcd eps_inv_j =
+          rpa.calculate_epsilon_complex(0.0, newpoint).inverse();
+      eps_inv_j.diagonal().array() -= 1.0;
+      _dielinv_matrices_r[j] = -eps_inv_j;
+    } else {
+      double newpoint = _quadpoints(j);
+      Eigen::MatrixXcd eps_inv_j =
+          rpa.calculate_epsilon_complex(0.0, newpoint).inverse();
+      eps_inv_j.diagonal().array() -= 1.0;
+      _dielinv_matrices_r[j] = -eps_inv_j;
+    }
+  }
+}
+
+double GaussianQuadrature::SigmaGQDiag(double frequency, Index gw_level,
+                                       double eta) const {
+  Index homo = _opt.homo - _opt.rpamin;
+  Index lumo = homo + 1;
+
+  const Eigen::MatrixXd& Imx = _Mmn[gw_level];
+  Eigen::ArrayXcd DeltaE = frequency - _energies.array();
+  std::complex<double> result = std::complex<double>(0.0, 0.0);
+  if (_opt.quadrature_scheme == "laguerre") {
+    // This work for integration limits a = 0 b = +infty
```

Comment:
What is `this`?

Reply:
for 'this' I meant the laguerre method. I will be more explicit